SSA/Projects/CIRCT/comb/comb.lean

```diff
@@ -0,0 +1,539 @@
+import SSA.Core.MLIRSyntax.EDSL
+import SSA.Projects.CIRCT.Stream.Stream
+import SSA.Projects.CIRCT.Stream.WeakBisim
+
+
+namespace Comb
+
+
+def add {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.add (0#w) l
+
+def and {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.and (0#w) l
+
+/-- Concatenate a list of bitvecs `xs`, where the length of bitvec xs[i] is given by
+  element ls[i] in a list of nat `ls`-/
+def concat {ls : List Nat} (xs : HVector BitVec ls) : BitVec (List.sum ls) :=
+  match (xs) with
+  | (.nil) => 0#0
+  | (.cons x xs) =>
+     let xsSum := concat xs
+     x ++ xsSum
+
+def divs (x y : BitVec w) : BitVec w :=
+  BitVec.sdiv x y
+
+def divu (x y : BitVec w) : BitVec w :=
+  BitVec.udiv x y
+
+def extract (x : BitVec w) (lb : Nat) : BitVec (w - lb) :=
+  BitVec.truncate (w - lb) (BitVec.ushiftRight x lb)
+
+def icmp (x y : BitVec w) (op : Nat) : Bool :=
+  match op with
```

Comment:
Defining `op` as a `Nat` is a big no-no: you should use an enum/inductive type for this. See [how the LLVM dialect](https://github.com/opencompl/lean-mlir/blob/4ffef329226f4f22004389b30f695ed8dc206301/SSA/Projects/InstCombine/LLVM/Semantics.lean#L416) does this for inspiration